src/server/analytic_services/prometheus_reports/noobaa_endpoint_report.js

```diff
@@ -2,38 +2,114 @@
+        type: 'Counter',
+        name: 'cache_write_bytes',
+        configuration: {
+            help: 'Cache write bytes in namespace cache bucket',
+            labelNames: ['namespace_resource_id']
+        }
+    },
+    {
+        type: 'Counter',
+        name: 'cache_object_read_count',
+        configuration: {
+            help: 'Counter on entire object reads in namespace cache bucket',
+            labelNames: ['namespace_resource_id']
+        }
+    },
+    {
+        type: 'Counter',
+        name: 'cache_object_read_miss_count',
+        configuration: {
+            help: 'Counter on entire object read miss in namespace cache bucket',
+            labelNames: ['namespace_resource_id']
+        }
+    },
+    {
+        type: 'Counter',
+        name: 'cache_range_read_count',
+        configuration: {
+            help: 'Counter on range reads in namespace cache bucket',
+            labelNames: ['namespace_resource_id']
+        }
+    },
+    {
+        type: 'Counter',
+        name: 'cache_range_read_miss_count',
+        configuration: {
+            help: 'Counter on range read miss in namespace cache bucket',
+            labelNames: ['namespace_resource_id']
+        }
+    },
+]);
+
 class NooBaaEndpointReport extends BasePrometheusReport {
     constructor() {
         super();
 
         if (this.enabled) {
-            // Cache the client for ease of use.
-            const client = this.prom_client;
-
-            // Define storage for the metrics:
-            this._metrics = {
-                // This is a dummy metric that is used as an example, Gauge is use
-                // here forbut it can be any of the metric types defined in the prom
-                // client lib at https://github.com/siimon/prom-client with its
-                // specific configuration.
-                demo_metric: new client.Gauge({
-                    name: this.get_prefixed_name('demo_metric'),
+            this._metrics = {};
+            for (const m of NOBBAA_ENDPOINT_METRICS) {
+                this._metrics[m.name] = new this.prom_client[m.type]({
+                    name: this.get_prefixed_name(m.name),
                     registers: [this.registry],
-                    help: 'Description line for demo_metric',
-                    // Other configurqation that should apply for this metric
-                }),
-            };
+                    ...m.configuration,
+                });
+            }
+        }
+
+        for (const { type, name } of NOBBAA_ENDPOINT_METRICS) {
+            if (type === 'Counter') {
+                this[`set_${name}`] = (...args) => {
+                    if (!this._metrics) return;
+                    this._metrics[name].inc(...args);
+                };
+            } else if (type === 'Histogram') {
+                this[`set_${name}`] = (...args) => {
+                    if (!this._metrics) return;
+                    this._metrics[name].observe(...args);
+                };
```

Comment:
Creating a function dynamically on a class instance will introduce some problems:
1. The js linter will not be aware of these functions.
2. Reading the class does not expose the class API, the API is hidden inside the constructor implantation
3. We may find in the future that we have some "special cases" that does not fall into these two categories, for example, a new type of metric or even worse, we would like to treat the setting of two or more metrics together as one call. Which leaves us with 2 options, either we make the generating code much more complex or we start to add some methods to the class, but in that case, we just split the definition of the API of the class into two separate locations and the reader will just get even more confused.

This is exactly the case we currently have with the `noobaa-core` report, which is bad, and I think we should not repeat the same mistake twice.

My approach would be as follows (the example is for `cache_range_read_count` but should be applied to all metrics):
In the constructor:
```
this._metrics['cache_range_read_count'] = new client.Counter({
  name: this.get_prefixed_name('cache_range_read_count'),
  registers: [this.registry],
  help: `Counter on range reads in namespace cache bucket`,
  labelNames: ['namespace_resource_id'] 
})
```

And one or more class methods that exposes the metric:
```
inc_cache_range_read_count(value) {
    if (!this._metrics) return;
    this._metrics['inc_cache_range_read_count'].inc(value)
}
```

I know it is more verbos, but in my eye this is exactly what er need it to be without adding more unnececary complexity

Reply:
I agree with 1 and 2. Regarding 3, I don't think supporting different types of metrics will cause generated functions to be complex. They all fall into similar prototype, set(labels, value), inc(labels, value), observe(labels, value) etc. 

Regarding your proposal manually adding individual class method for each metric, the class methods would have similar code. It would be error prone in terms of coding, if programmer tried to copy and paste to create class method for new metric, and forgot to change the name of the metric in (this._metrics['wrong_name']), it would be annoying issue to figure it out. If there are too many metrics, it would just too boring to write. It is why we tend to create generic function to make code concise and easy to maintain.

If you don't think creating class methods during run-time is good idea, would you be ok with different way of creating generic function?, such as 
```
function observe(metric_name, labels, value) {
    if (!this.metrics) return;
    const metric = this._metrics['inc_cache_range_read_count'];
    if (!metric) throw new Error(...'');
    if (metric.type === 'Counter') metric.inc(labels, value);
    else if if (metric.type === 'Histogram') metric.observe(labels, value);
    else throw new Error('...');
}
```

Reply:
To my eyes, this seems much easier to understand than the original version as it defines a clean interface to the class (althougth a generic one). 

But let's consider the points you have made:
1. Yes the code will be similar, but it will be small and readable. The linter and autocomplete features of our IDEs will help us when using the class as the interface is fully defined. So trying to update a non existant metric will be catched immediately. 

2. I agree that while coding you might make errors while writing new accessor methods, but consider that these methods are only written once, while errors in passing a wrong name for a metric in call sites has greater possiblity of occurence. 
These errors can be cached only in runtime and only when these code paths are active.
Let's say tomorrow you would like to rename a metric and it's accessor, in my proposal, as soon as you change the method name the linter will complain on all possible call sites. In your proposal we will have to find any potantial bug in runtime. Also, the number of metrics is finite which means that this effort should only happen once.

3. Sometimes code is boarding, this is the nature of the beast.

4. You skipped the biggest point on my list: It might happen that you will find that some code paths need to update one metric based on another or even conditionally. In that case, you would like to capture this logic inside the report class in order to avoid duplication in all call sites (you can see an example of this happening to us with the core's report). 
The cleanest way to do that is to create a non-naive setter that capture that logic, trying to describe this using a data model is hard and over complicating the entire process. In addition to that, you will not want to give access to the actual metrics behind the scenes so you will now have logic that skips the creation of auto setters for these metrics which again make it even more complex.

Saying all of the above I still think a verbose interface is the cleanest, with minimal complexity, more readable, easier to change,  and will do the job just fine.

If you still want to do something else, I will suggest not having one `observe` methods but one for each metric kind: 
` inc(name, labels, value)`
`set(name, value)`
`observe(labels, value)`

At the minimum, this solution will allow the reader of the call site to understand the updated pattern.

Reply:
@nb-ohad 
Regading 2, if calling function passes wrong name, it would be easy to catch since exception on unknown metric would be thrown. Let's say developer creates new metric function by copying and pasting existing one, and he/she only changes the function name and forgets to change the metric name inside the function, you will still have problem in runtime, and it is annoying to find the cause, since you would have two metrics having wrong value. In short, I see both methods have pros and cons. Probably not all will agree on same one.

Regarding 4, in the case that logic of updating metric is complex, we will create customize function on top of the basic ones (e.g, inc, set, observe). I am not against creating such function. It is what we have to do anyway.

Reply:
@paulxuzhang I get your point, and I understand the pros and cons. 
And as I said before, it is just my opinion, coming from my past experience.

So let's agree to do your suggestion, the one from the first comment, with or without the separation to `inc/set/observe` to your choosing. sounds ok?